// @taquito/tzip12/dist/lib/tzip12-contract-abstraction.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tzip12ContractAbstraction = void 0;
var michelson_encoder_1 = require("@taquito/michelson-encoder");
var tzip16_1 = require("@taquito/tzip16");
var tzip12_errors_1 = require("./tzip12-errors");
var tokenMetadataBigMapType = {
    prim: 'big_map',
    args: [
        { prim: 'nat' },
        { prim: 'pair', args: [
                { prim: 'nat', annots: ['%token_id'] },
                { prim: "map", args: [{ prim: 'string' }, { prim: 'bytes' }], annots: ['%token_info'] }
            ] }
    ],
    annots: ['%token_metadata']
};
var Tzip12ContractAbstraction = /** @class */ (function () {
    function Tzip12ContractAbstraction(contractAbstraction, context) {
        this.contractAbstraction = contractAbstraction;
        this.context = context;
        this._tzip16ContractAbstraction = new tzip16_1.Tzip16ContractAbstraction(this.contractAbstraction, this.context);
    }
    /**
     * @description Fetches the contract metadata (according to the Tzip-016 standard)
     * @returns An object containing the metadata, the uri, an optional integrity check result and an optional sha256 hash
     * or `Undefined` if the contract has no metadata (non-compliant with Tzip-016)
     */
    Tzip12ContractAbstraction.prototype.getContractMetadata = function () {
        return __awaiter(this, void 0, void 0, function () {
            var contractMetadata, err_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this._tzip16ContractAbstraction.getMetadata()];
                    case 1:
                        contractMetadata = _a.sent();
                        return [2 /*return*/, contractMetadata.metadata];
                    case 2:
                        err_1 = _a.sent();
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @description The Tzip-016 "interfaces" field MUST be present in the contract metadata. It should contain "TZIP-012[version-info]"
     * @returns True if "interfaces" field is present and contains "TZIP-012", false otherwise
     */
    Tzip12ContractAbstraction.prototype.isTzip12Compliant = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var isCompliant, metadata, tzip12Interface;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        isCompliant = false;
                        return [4 /*yield*/, this.getContractMetadata()];
                    case 1:
                        metadata = _b.sent();
                        if (metadata) {
                            tzip12Interface = (_a = metadata.interfaces) === null || _a === void 0 ? void 0 : _a.filter(function (x) {
                                return x.substring(0, 8) === "TZIP-012";
                            });
                            isCompliant = (tzip12Interface && tzip12Interface.length !== 0) ? true : false;
                        }
                        return [2 /*return*/, isCompliant];
                }
            });
        });
    };
    /**
     * @description Fetches the token metadata for a specified token ID.
     * The function first tries to find a `token_metadata` view in the contract metadata and to execute it with the token ID.
     * If there is no view, the function tries to find a `token_metadata` bigmap in the top-level pairs of the storage.
     * @param tokenId The ID of the token for which we want to retrieve token metadata
     * @returns An object of type `TokenMetadata`
     */
    Tzip12ContractAbstraction.prototype.getTokenMetadata = function (tokenId) {
        return __awaiter(this, void 0, void 0, function () {
            var tokenMetadata;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.retrieveTokenMetadataFromView(tokenId)];
                    case 1:
                        tokenMetadata = _a.sent();
                        return [2 /*return*/, (!tokenMetadata) ? this.retrieveTokenMetadataFromBigMap(tokenId) : tokenMetadata];
                }
            });
        });
    };
    Tzip12ContractAbstraction.prototype.retrieveTokenMetadataFromView = function (tokenId) {
        return __awaiter(this, void 0, void 0, function () {
            var views;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getContractMetadata()];
                    case 1:
                        if (!_a.sent()) return [3 /*break*/, 3];
                        return [4 /*yield*/, this._tzip16ContractAbstraction.metadataViews()];
                    case 2:
                        views = _a.sent();
                        if (views && this.hasTokenMetadataView(views)) {
                            return [2 /*return*/, this.executeTokenMetadataView(views['token_metadata'](), tokenId)];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    Tzip12ContractAbstraction.prototype.hasTokenMetadataView = function (views) {
        var e_1, _a;
        try {
            for (var _b = __values(Object.keys(views)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var view = _c.value;
                if (view === 'token_metadata') {
                    return true;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return false;
    };
    Tzip12ContractAbstraction.prototype.executeTokenMetadataView = function (tokenMetadataView, tokenId) {
        return __awaiter(this, void 0, void 0, function () {
            var tokenMetadata, err_2, tokenMap, metadataFromUri;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, tokenMetadataView.executeView(tokenId)];
                    case 1:
                        tokenMetadata = _a.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        err_2 = _a.sent();
                        throw new tzip12_errors_1.TokenIdNotFound(tokenId);
                    case 3:
                        tokenMap = Object.values(tokenMetadata)[1];
                        if (!michelson_encoder_1.MichelsonMap.isMichelsonMap(tokenMap)) {
                            throw new tzip12_errors_1.TokenMetadataNotFound(this.contractAbstraction.address);
                        }
                        return [4 /*yield*/, this.fetchTokenMetadataFromUri(tokenMap)];
                    case 4:
                        metadataFromUri = _a.sent();
                        return [2 /*return*/, this.formatMetadataToken(tokenId, tokenMap, metadataFromUri)];
                }
            });
        });
    };
    Tzip12ContractAbstraction.prototype.fetchTokenMetadataFromUri = function (tokenMetadata) {
        return __awaiter(this, void 0, void 0, function () {
            var uri, metadataFromUri, e_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        uri = tokenMetadata.get("");
                        if (!uri) return [3 /*break*/, 4];
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.context.metadataProvider.provideMetadata(this.contractAbstraction, tzip16_1.bytes2Char(uri), this.context)];
                    case 2:
                        metadataFromUri = _a.sent();
                        return [2 /*return*/, metadataFromUri.metadata];
                    case 3:
                        e_2 = _a.sent();
                        if (e_2.name === 'InvalidUri') {
                            console.warn("The URI " + tzip16_1.bytes2Char(uri) + " is present in the token metadata, but is invalid.");
                        }
                        else {
                            throw e_2;
                        }
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    Tzip12ContractAbstraction.prototype.formatMetadataToken = function (tokenId, metadataTokenMap, metadataFromUri) {
        var e_3, _a, _b, _c, _d;
        var tokenMetadataDecoded = {
            'token_id': tokenId
        };
        try {
            for (var _e = __values(metadataTokenMap.keys()), _f = _e.next(); !_f.done; _f = _e.next()) {
                var keyTokenMetadata = _f.value;
                if (keyTokenMetadata === 'decimals') {
                    Object.assign(tokenMetadataDecoded, (_b = {}, _b[keyTokenMetadata] = Number(tzip16_1.bytes2Char(metadataTokenMap.get(keyTokenMetadata))), _b));
                }
                else if (!(keyTokenMetadata === '')) {
                    Object.assign(tokenMetadataDecoded, (_c = {}, _c[keyTokenMetadata] = tzip16_1.bytes2Char(metadataTokenMap.get(keyTokenMetadata)), _c));
                }
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_f && !_f.done && (_a = _e.return)) _a.call(_e);
            }
            finally { if (e_3) throw e_3.error; }
        }
        // if an URI is present, add the fetched properties to the object
        // if a property is in the URI and the map, prevalence is accorded to value from the URI
        if (metadataFromUri) {
            for (var property in metadataFromUri) {
                Object.assign(tokenMetadataDecoded, (_d = {}, _d[property] = metadataFromUri[property], _d));
            }
        }
        if (!('decimals' in tokenMetadataDecoded)) {
            throw new tzip12_errors_1.InvalidTokenMetadata();
        }
        return tokenMetadataDecoded;
    };
    Tzip12ContractAbstraction.prototype.retrieveTokenMetadataFromBigMap = function (tokenId) {
        return __awaiter(this, void 0, void 0, function () {
            var bigmapTokenMetadataId, pairNatMap, err_3, michelsonMap, metadataFromUri;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        bigmapTokenMetadataId = this.findTokenMetadataBigMap();
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.context.contract.getBigMapKeyByID(bigmapTokenMetadataId['int'].toString(), tokenId.toString(), new michelson_encoder_1.Schema(tokenMetadataBigMapType))];
                    case 2:
                        pairNatMap = _a.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        err_3 = _a.sent();
                        throw new tzip12_errors_1.TokenIdNotFound(tokenId);
                    case 4:
                        michelsonMap = pairNatMap['token_info'];
                        if (!michelson_encoder_1.MichelsonMap.isMichelsonMap(michelsonMap)) {
                            throw new tzip12_errors_1.TokenIdNotFound(tokenId);
                        }
                        return [4 /*yield*/, this.fetchTokenMetadataFromUri(michelsonMap)];
                    case 5:
                        metadataFromUri = _a.sent();
                        return [2 /*return*/, this.formatMetadataToken(tokenId, michelsonMap, metadataFromUri)];
                }
            });
        });
    };
    Tzip12ContractAbstraction.prototype.findTokenMetadataBigMap = function () {
        var tokenMetadataBigMapId = this.contractAbstraction.schema.FindFirstInTopLevelPair(this.contractAbstraction.script.storage, tokenMetadataBigMapType);
        if (!tokenMetadataBigMapId) {
            throw new tzip12_errors_1.TokenMetadataNotFound(this.contractAbstraction.address);
        }
        return tokenMetadataBigMapId;
    };
    return Tzip12ContractAbstraction;
}());
exports.Tzip12ContractAbstraction = Tzip12ContractAbstraction;
