// @taquito/tzip12/dist/lib/tzip12-errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidTokenMetadata = exports.TokenIdNotFound = exports.TokenMetadataNotFound = void 0;
var TokenMetadataNotFound = /** @class */ (function () {
    function TokenMetadataNotFound(address) {
        this.address = address;
        this.name = 'TokenMetadataNotFound';
        this.message = "No token metadata was found for the contract: " + address;
    }
    return TokenMetadataNotFound;
}());
exports.TokenMetadataNotFound = TokenMetadataNotFound;
var TokenIdNotFound = /** @class */ (function () {
    function TokenIdNotFound(tokenId) {
        this.tokenId = tokenId;
        this.name = 'TokenIdNotFound';
        this.message = "Could not find token metadata for the token ID: " + tokenId;
    }
    return TokenIdNotFound;
}());
exports.TokenIdNotFound = TokenIdNotFound;
var InvalidTokenMetadata = /** @class */ (function () {
    function InvalidTokenMetadata() {
        this.name = 'InvalidTokenMetadata';
        this.message = 'Non-compliance with the TZIP-012 standard. The required property `decimals` is missing.';
    }
    return InvalidTokenMetadata;
}());
exports.InvalidTokenMetadata = InvalidTokenMetadata;
